src/cls/queue/cls_queue.cc

```diff
@@ -0,0 +1,1257 @@
+      return -EINVAL;
+    }
+  }
+
+  bool is_last_entry = false;
+  in->clear();
+  out->clear();
+  ret = cls_queue_get_last_entry(hctx, in, out);
+  if (ret < 0) {
+    return ret;
+  }
+
+  cls_rgw_gc_obj_info info;
+  auto iter = out->cbegin();
+  try {
+    decode(info, iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_gc_queue_update_entry(): failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  CLS_LOG(1, "INFO: tag of gc info is %s\n", info.tag.c_str());
+  if (info.tag == op.info.tag) {
+    is_last_entry = true;
+  }
+
+  cls_enqueue_op enqueue_op;
+  encode(op.info, enqueue_op.data.bl_data);
+  enqueue_op.data.size_data = enqueue_op.data.bl_data.length();
+  CLS_LOG(1, "INFO: cls_gc_update_entry: Data size is: %lu \n", enqueue_op.data.size_data);
+
+  auto it = urgent_data_map.find(op.info.tag);
+  if (it != urgent_data_map.end()) {
+    it->second = op.info.time;
+  } else {
+    urgent_data_map.insert({op.info.tag, op.info.time});
+    enqueue_op.has_urgent_data = true;
+  }
+
+  encode(urgent_data_map, enqueue_op.bl_urgent_data);
+
+  out->clear();
+  bool can_fit = false;
+  ret = cls_queue_can_urgent_data_fit(hctx, &(enqueue_op.bl_urgent_data), out);
+  if (ret < 0) {
+     return ret;
+  }
+  iter = out->cbegin();
+  decode(can_fit, iter);
+  CLS_LOG(1, "INFO: Can urgent data fit: %d \n", can_fit);
+
+  if (can_fit) {
+    in->clear();
+    encode(enqueue_op, *in);
+    if (! is_last_entry) {
+      ret = cls_enqueue(hctx, in, out);
+      if (ret < 0) {
+        return ret;
+      }
+    } else {
+      ret = cls_queue_update_last_entry(hctx, in, out);
+      if (ret < 0) {
+        return ret;
+      }
+    }
+  }
+  // Else write urgent data to omap as xattrs
+  else {
+    bufferlist bl_time;
+    encode(op.info.tag, bl_time);
+    ret = cls_cxx_setxattr(hctx, op.info.tag.c_str(), &bl_time);
+    CLS_LOG(20, "%s(): setting attr: %s", __func__, op.info.tag.c_str());
+    if (ret < 0) {
+      CLS_LOG(0, "ERROR: %s(): cls_cxx_setxattr (attr=%s) returned %d", __func__, op.info.tag.c_str(), ret);
+      return ret;
+    }
+  }
+  return 0;
+}
+
+CLS_INIT(queue)
+{
+  CLS_LOG(1, "Loaded queue class!");
+
+  cls_handle_t h_class;
+  cls_method_handle_t h_gc_create_queue;
+  cls_method_handle_t h_get_queue_size;
+  cls_method_handle_t h_gc_enqueue;
+  cls_method_handle_t h_gc_dequeue;
+  cls_method_handle_t h_gc_queue_list;
+  cls_method_handle_t h_gc_queue_remove;
+  cls_method_handle_t h_gc_queue_update;
+
+  cls_register(QUEUE_CLASS, &h_class);
+
+  /* queue*/
+  cls_register_cxx_method(h_class, GC_CREATE_QUEUE, CLS_METHOD_RD | CLS_METHOD_WR, cls_gc_create_queue, &h_gc_create_queue);
+  cls_register_cxx_method(h_class, GET_QUEUE_SIZE, CLS_METHOD_RD | CLS_METHOD_WR, cls_get_queue_size, &h_get_queue_size);
+  cls_register_cxx_method(h_class, GC_ENQUEUE, CLS_METHOD_RD | CLS_METHOD_WR, cls_gc_enqueue, &h_gc_enqueue);
```

Comment:
I am thinking of modifying cls_enqueue to take a list of entries instead of a single entry. The client can pass a single entry if it wants to. Is this ok?